src/steth/scWETHv2Keeper.sol

```diff
@@ -0,0 +1,266 @@
+        }
+
+        target.rebalance(_calculateInvestAmount(), _flashLoanAmount, _multicallData);
+
+        uint256 wstEthBalance = wstEth.balanceOf(address(target));
+
+        if (wstEthBalance != 0) target.supplyAndBorrow(_adapterIdForLeftoverWstEth, wstEthBalance, 1);
+    }
+
+    /**
+     * @notice Calculates the investment parameters for rebalancing.
+     * @dev This function performs various checks on the input parameters and calculates the necessary amounts for rebalancing.
+     * @param _adapterIds The array of adapter IDs.
+     * @param _allocations The array of allocations corresponding to each adapter ID.
+     * @param _targetLtvs The array of target LTVs corresponding to each adapter ID.
+     * @return totalInvestAmount The total amount to invest.
+     * @return totalFlashLoanAmount The total amount to flashloan.
+     * @return rebalanceMulticallData The array of encoded function calls for the rebalance operation.
+     *
+     * @custom:requirements
+     * - `_adapterIds`, `_allocations`, and `_targetLtvs` must have the same length and must not be empty.
+     * - `_allocations` must sum to one.
+     * - Each `_allocation` must be non-zero.
+     * - Each `_targetLtv` must be non-zero.
+     * - Each `_adapterId` must be supported by the target contract.
+     *
+     * @custom:reverts
+     * - `NoNeedToInvest` if the current float is less than the minimum required float.
+     * - `InvalidInputParameters` if the lengths of the input arrays are not equal or are zero.
+     * - `ZeroAllocation` if any of the allocations is zero.
+     * - `AllocationsMustSumToOne` if the total allocation does not sum to one.
+     * - `ProtocolNotSupported` if any of the specified adapter IDs is not supported.
+     * - `ZeroTargetLtv` if any of the target LTVs is zero.
+     */
+    function calculateInvestParams(
+        uint256[] calldata _adapterIds,
+        uint256[] calldata _allocations,
+        uint256[] calldata _targetLtvs
+    )
+        external
+        view
+        returns (uint256 totalInvestAmount, uint256 totalFlashLoanAmount, bytes[] memory rebalanceMulticallData)
+    {
+        if (
+            _adapterIds.length == 0 || _adapterIds.length != _allocations.length
+                || _adapterIds.length != _targetLtvs.length
+        ) {
+            revert InvalidInputParameters();
+        }
+
+        uint256 totalAllocation;
+        for (uint256 i = 0; i < _allocations.length; i++) {
+            if (_allocations[i] == 0) revert ZeroAllocation();
+            totalAllocation += _allocations[i];
+        }
+
+        if (totalAllocation != C.ONE) revert AllocationsMustSumToOne();
+
+        rebalanceMulticallData = new bytes[](_adapterIds.length + 1); // +1 for swapWethToWstEth
+        // start at 1 because the first call is for swapping WETH to wstETH
+        uint256 callDataIndex = 1;
+        totalInvestAmount = _calculateInvestAmount();
+
+        for (uint256 i = 0; i < _adapterIds.length; i++) {
```

Comment:
should define i outside of loop right? :innocent: and ++i ?

Reply:
not really, it doesn't make a difference where the variable i is defined. i++ or ++i is not that relevant in this case (even tho ++i is a little more gas efficient) since the function was intended to be called off-chain only, because of all of the computation and memory allocation logic it holds. but thanks for pointing that out, it made me realize i haven't documented that this function should ideally be called off-chain only